src/mon/ElectionLogic.h

```diff
@@ -0,0 +1,349 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software 
+ * Foundation.  See file COPYING.
+ * 
+ */
+
+
+#ifndef CEPH_ELECTIONLOGIC_H
+#define CEPH_ELECTIONLOGIC_H
+
+#include <map>
+#include "include/types.h"
+
+class ElectionOwner {
+public:
+  /**
+   * Write down the given epoch in persistent storage, such that it
+   * can later be retrieved by read_persisted_epoch even across process
+   * or machine restarts.
+   *
+   * @param e The epoch to write
+   */
+  virtual void persist_epoch(epoch_t e) = 0;
+  /**
+   * Retrieve the most-previously-persisted epoch.
+   *
+   * @returns The latest epoch passed to persist_epoch()
+   */
+  virtual epoch_t read_persisted_epoch() = 0;
+  /**
+   * Validate that the persistent store is working by committing
+   * to it. (There is no interface for retrieving the value; this
+   * tests local functionality before doing things like triggering
+   * elections to try and join a quorum.)
+   */
+  virtual void validate_store() = 0;
+  /**
+   * Notify the ElectionOwner that ElectionLogic has increased its
+   * election epoch. This resets an election (either on local loss or victory,
+   * or when trying a new election round) and the ElectionOwner
+   * should reset any tracking of its own to match. (The ElectionLogic
+   * will further trigger sending election messages if that is
+   * appropriate.)
+   */
+  virtual void notify_bump_epoch() = 0;
+  /**
+   * Notify the ElectionOwner we must start a new election.
+   */
+  virtual void trigger_new_election() = 0;
+  /**
+   * Retrieve this Paxos instance's rank.
+   */
+  virtual int get_my_rank() = 0;
+  /**
+   * Send a PROPOSE message to all our peers. This happens when
+   * we have started a new election (which may mean attempting to
+   * override a current one).
+   *
+   * @param e The election epoch of our proposal.
+   */
+  virtual void propose_to_peers(epoch_t e) = 0;
+  /**
+   * The election has failed and we aren't sure what the state of the
+   * quorum is, so reset the entire system as if from scratch.
+   */
+  virtual void reset_election() = 0;
+  /**
+   * Ask the ElectionOwner if we-the-Monitor have ever participated in the
+   * quorum (including across process restarts!).
+   *
+   * @returns true if we have participated, false otherwise
+   */
+  virtual bool ever_participated() = 0;
+  /**
+   * Ask the ElectionOwner for the size of the Paxos set. This includes
+   * those monitors which may not be in the current quorum!
+   */
+  virtual unsigned paxos_size() = 0;
+  /**
+   * Tell the ElectionOwner we have started a new election.
+   *
+   * The ElectionOwner is responsible for timing out the election (by invoking
+   * end_election_period()) if it takes too long (as defined by the ElectionOwner).
+   * This function is the opportunity to do that and to clean up any other external 
+   * election state it may be maintaining.
+   */
+  virtual void _start() = 0;
```

Comment:
why the name starts with `_` while it's public?

Reply:
It's a backwards call from ElectionLogic into ElectionOwner, rather than something others should be invoking. Plus it gets called from ElectionOwner::start().
It's not perfect naming but it's what I could come up with.